# Task 1 CNN for Text Classification

# Introduction

This part is task 1 of HW3 focusing on the effect of NLP features in CNN for Text Classification. We developed a CNN for the text classification NLP task and most importantly concatinating NLP feature and word embedding as a new feature.


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
from sklearn.utils import shuffle

In [3]:
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
from keras.models import Sequential
from keras.optimizers import SGD
from keras.layers import Embedding, Conv1D, Dropout, MaxPooling1D, Flatten, Dense
from keras import regularizers

Using TensorFlow backend.


In [29]:
from keras.models import Model
from keras.optimizers import SGD
from keras.layers import Input, Embedding, Conv1D, Dropout, MaxPooling1D, Flatten, Dense, concatenate
from keras import regularizers


In [4]:
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import StratifiedKFold, cross_val_score
from keras.callbacks import EarlyStopping
from sklearn.metrics import accuracy_score

## data loading and preprocessing

In [5]:
df_n = pd.read_table('/Users/jianwenliu/nlp/rt-polaritydata/rt-polaritydata/rt-polarity.neg', names=['review'],  header=None, encoding='latin-1')
df_p = pd.read_table('/Users/jianwenliu/nlp/rt-polaritydata/rt-polaritydata/rt-polarity.pos', names=['review'],  header=None, encoding='latin-1')


/Users/jianwenliu/anaconda3/envs/deepAI/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\t'.
  """Entry point for launching an IPython kernel.
/Users/jianwenliu/anaconda3/envs/deepAI/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\t'.
  


In [6]:
df_n.head(10)

review
0                  simplistic , silly and tedious . 
1  it's so laddish and juvenile , only teenage bo...
2  exploitative and largely devoid of the depth o...
3  [garbus] discards the potential for pathologic...
4  a visually flashy but narratively opaque and e...
5  the story is also as unoriginal as they come ,...
6  about the only thing to give the movie points ...
7                    not so much farcical as sour . 
8  unfortunately the story and the actors are ser...
9  all the more disquieting for its relatively go...

label the review with 0 and 1

In [7]:
df_n['lable'] = 0
df_p['lable'] = 1

Concate and shuffle the dataset, making it 50-50

In [8]:
df = shuffle(pd.concat((df_n, df_p), axis=0), random_state=7)

In [9]:
df

review  lable
1777  this is an egotistical endeavor from the daugh...      0
2272  where last time jokes flowed out of cho's life...      0
155   everything that has to do with yvan and charlo...      1
4137  an affectionately goofy satire that's unafraid...      1
2428  i was impressed by how many tit-for-tat retali...      1
3039  what [denis] accomplishes in his chilling , un...      1
2411  a journey that's too random and inconclusive t...      0
3626  promises is one film that's truly deserving of...      1
2463  all this turns out to be neither funny nor pro...      0
2302  the stripped-down dramatic constructs , auster...      1
2896  even as it pays earnest homage to turntablists...      1
2786  the scorpion king is more fun than conan the b...      1
1331  there is a fabric of complex ideas here , and ...      1
1685  in exactly 89 minutes , most of which passed a...      0
3236  too many scenarios in which the hero might hav...      0
571   a seriously bad film with seriously warped log...      0
4706  don't wait to see this terrific film with your...      1
1926  uplifting as only a document of the worst poss...      1
4164  the movie is like scorsese's mean streets redo...      0
4426  the pretensions -- and disposable story -- sin...      0
675   the re- enactments , however fascinating they ...      0
3352  be forewarned , if you're depressed about anyt...      0
5157  has enough wit , energy and geniality to pleas...      1
2443  matthew mcconaughey tries , and fails , to con...      0
349   over-the-top and a bit ostentatious , this is ...      1
4445  been there , done that . . . a thousand times ...      0
2021  even horror fans will most likely not find wha...      0
2759  one of those unassuming films that sneaks up o...      1
3357  daughter from danang reveals that efforts towa...      1
2237  a raunchy and frequently hilarious follow-up t...      1
...                                                 ...    ...
2320  if cinema had been around to capture the chaos...      1
4022                 a small movie with a big impact .       1
3633  [it's] a clever thriller with enough unexpecte...      1
251   one of the greatest romantic comedies of the p...      1
3173  if you ever wondered what it would be like to ...      1
4859  don michael paul uses quick-cuts , ( very ) la...      0
2997  flotsam in the sea of moviemaking , not big en...      0
1176  the only way to tolerate this insipid , brutal...      0
3896  there ain't a lot more painful than an unfunny...      0
4210  a treat for its depiction on not giving up on ...      1
3516  the filmmakers needed more emphasis on the sto...      0
2624  very well written and directed with brutal hon...      1
5159  bittersweet comedy/drama full of life , hand g...      1
2349  it ain't art , by a long shot , but unlike las...      1
1753  s1m0ne's satire is not subtle , but it is effe...      1
3252  the film is about the relationships rather tha...      1
684   the movie sticks much closer to hornby's drop-...      1
1456  seems like something american and european gay...      0
1092  gibney and jarecki just want to string the bas...      0
787   that rara avis : the intelligent romantic come...      1
3239  broomfield's style of journalism is hardly jou...      0
4441  showtime isn't particularly assaultive , but i...      0
2583  an overstuffed compendium of teen-catholic-mov...      0
2235  amazing ! a college story that works even with...      1
1372  so boring that even its target audience talked...      0
919   nothing more than an amiable but unfocused bag...      0
4307  every joke is repeated at least four times . e...      0
368   by the end of no such thing the audience , lik...      1
537   a 75-minute sample of puerile rubbish that is ...      0
4081  i'd watch these two together again in a new yo...      1

[10662 rows x 2 columns]

In [10]:
df.shape

(10662, 2)

In [11]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts(df['review'])

In [12]:
sequences = tokenizer.texts_to_sequences(df['review'])

In [13]:
max_len = max([len(seq) for seq in sequences])
max_len

51

Padding the sequences

In [14]:
sequences = pad_sequences(sequences, maxlen=max_len, dtype='int32', padding='pre', truncating='pre', value=0)

In [15]:
sequences

array([[   0,    0,    0, ...,    9, 1022, 4625],
       [   0,    0,    0, ..., 1074, 1023, 3890],
       [   0,    0,    0, ...,   64,    3,  281],
       ...,
       [   0,    0,    0, ...,   14,    1,  941],
       [   0,    0,    0, ...,  282, 4424,  147],
       [   0,    0,    0, ...,   93, 1003,  369]], dtype=int32)

get thr dictionaty for the word index

In [16]:
tokenizer.word_index

{'the': 1,
 'a': 2,
 'and': 3,
 'of': 4,
 'to': 5,
 'is': 6,
 'in': 7,
 'that': 8,
 'it': 9,
 'as': 10,
 'but': 11,
 'with': 12,
 'film': 13,
 'for': 14,
 'this': 15,
 'its': 16,
 'an': 17,
 'movie': 18,
 "it's": 19,
 'be': 20,
 'on': 21,
 'you': 22,
 'not': 23,
 'by': 24,
 'one': 25,
 'like': 26,
 'about': 27,
 'more': 28,
 'has': 29,
 'are': 30,
 'at': 31,
 'than': 32,
 'from': 33,
 'all': 34,
 'his': 35,
 'have': 36,
 'so': 37,
 'if': 38,
 'or': 39,
 'story': 40,
 'too': 41,
 'i': 42,
 'out': 43,
 'just': 44,
 'who': 45,
 'up': 46,
 'good': 47,
 'into': 48,
 'what': 49,
 'most': 50,
 'no': 51,
 'much': 52,
 'even': 53,
 'comedy': 54,
 'time': 55,
 'will': 56,
 'can': 57,
 'some': 58,
 'well': 59,
 'characters': 60,
 'only': 61,
 'little': 62,
 'way': 63,
 'funny': 64,
 'their': 65,
 'director': 66,
 'make': 67,
 'been': 68,
 'your': 69,
 'enough': 70,
 'very': 71,
 'never': 72,
 'when': 73,
 'there': 74,
 'makes': 75,
 'life': 76,
 'bad': 77,
 'may': 78,
 'which': 79,
 'us': 80,
 'b

get the voc size

In [17]:
voc_size=len(tokenizer.word_index)

In [18]:
voc_size

19498

In [20]:
import spacy

In [22]:
nlp = spacy.load('en')

# create list to track tags for each review
tag_list = list()
tag_idx_list = list()

# create dictionaries to track vacabulary indexes
voc_tag = dict()
voc_tag_cnt = 0

for review in df.review.tolist():
    comm_tag_list = list()
    comm_tag_idx_list = list()
    doc = nlp(review)
    for token in doc:
        w = token.text
        if str.strip(w) not in ['','.']:      
            # tag int (fine pos tag)
            g = token.tag_
            if voc_tag.get(g,voc_tag_cnt) == voc_tag_cnt:
                voc_tag[g] = voc_tag_cnt 
                voc_tag_cnt += 1
            comm_tag_list.append(g)           
            comm_tag_idx_list.append(voc_tag[g]) 
    tag_list.append(comm_tag_list)
    tag_idx_list.append(comm_tag_idx_list)



In [23]:
X_tag = np.array(tag_list)
X_tag_idx = np.array(tag_idx_list)

In [24]:
X_tag.shape

(10662,)

pad sequences to for tags

In [25]:
tag_seqs = pad_sequences(X_tag_idx, maxlen=max_len, dtype='int32', padding='pre', truncating='pre', value=0.0)

In [26]:
voc_size_tag = len(voc_tag)

In [27]:
voc_size_tag

48

## CNN model

Build the CNN model

In [30]:
def conv_model(output_dim=64, filter_size=128, window_size=3, stride=1, pool_size=2, dense_dim=16):
    
    # for word
    krn_size = output_dim * window_size
    input1 = Input(shape=(max_len,), dtype='int32')
    embed1 = Embedding(voc_size+1, output_dim, input_length=max_len, embeddings_initializer='random_uniform')(input1)
    C1 = Conv1D(filter_size, kernel_size=krn_size ,padding='same', strides=(stride), activation='relu')(embed1)
    C1 = Dropout(0.5)(C1)
    M1 = MaxPooling1D(pool_size=(pool_size), padding='same')(C1)
    F1 = Flatten()(M1)
    D1 = Dense(dense_dim, activation='relu')(F1)
    D1 = Dropout(0.5)(D1)
    
    # for POS tag
    input2 = Input(shape=(max_len,), dtype='int32')
    embed2 = Embedding(voc_size_tag+1, int(output_dim/2), input_length=max_len, \
                       embeddings_initializer='random_uniform')(input2)
    C2 = Conv1D(int(filter_size/2), kernel_size=int(krn_size/2) ,padding='same', \
                       strides=(stride), activation='relu')(embed2)
    C2 = Dropout(0.5)(C2)
    M2 = MaxPooling1D(pool_size=(pool_size), padding='same')(C2)
    F2 = Flatten()(M2)
    D2 = Dense(1, activation='relu')(F2)
    D2 = Dropout(0.5)(D2)
    
    # Merge branches by concatenating along axis 1
    L1 = concatenate([D1, D2], axis=1)
    G1= Dense(dense_dim, activation='relu')(L1)
    G1 = Dropout(0.5)(G1)

    pred = Dense(1, kernel_regularizer=regularizers.l2(0.01),\
                 activity_regularizer=regularizers.l1(0.01),\
                 activation='sigmoid')(G1)
    
    model = Model(inputs=[input1, input2], outputs=pred)
    
    #sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    
    return model
    
conv_model().summary()

__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLayer)            (None, 51)           0                                            
__________________________________________________________________________________________________
input_2 (InputLayer)            (None, 51)           0                                            
__________________________________________________________________________________________________
embedding_1 (Embedding)         (None, 51, 64)       1247936     input_1[0][0]                    
__________________________________________________________________________________________________
embedding_2 (Embedding)         (None, 51, 32)       1568        input_2[0][0]                    
__________________________________________________________________________________________________
conv1d_1 (

In [32]:
EPOCHS = 10
BATCH_SIZE = 64
fold=8

# concatenates
X = np.hstack((sequences, tag_seqs))
y = np.array(df.lable.tolist())

K-Fold for validation

In [33]:
kfold = StratifiedKFold(n_splits=fold, shuffle=True, random_state=7)

## Model performance : prediction accuracy

In [34]:
acr = list()
i = 0
for train_index, valid_index in kfold.split(X, y):
    i += 1
    X_train, X_valid = X[train_index], X[valid_index]
    y_train, y_valid = y[train_index], y[valid_index]
    model = conv_model()
    # divide input back into sequences and tag_seqs
    model.fit([X_train[:,:max_len], X_train[:,max_len:]], y_train, batch_size=BATCH_SIZE, epochs=EPOCHS, verbose=2, \
              validation_data=([X_valid[:,:max_len], X_valid[:,max_len:]], y_valid),\
              callbacks=[EarlyStopping(patience=1, monitor='val_loss')])
    y_h = model.predict([X_valid[:,:max_len], X_valid[:,max_len:]])
    y_pred = [round(pred) for pred in y_h.reshape(-1)]
    
    acr.append(accuracy_score(y_valid, y_pred))

Train on 9328 samples, validate on 1334 samples
Epoch 1/10
 - 138s - loss: 0.9959 - acc: 0.5020 - val_loss: 0.9499 - val_acc: 0.5000
Epoch 2/10
 - 144s - loss: 0.9100 - acc: 0.5180 - val_loss: 0.8685 - val_acc: 0.6732
Epoch 3/10
 - 128s - loss: 0.7765 - acc: 0.6683 - val_loss: 0.8118 - val_acc: 0.7354
Epoch 4/10
 - 834s - loss: 0.6649 - acc: 0.7846 - val_loss: 0.8521 - val_acc: 0.7376
Train on 9328 samples, validate on 1334 samples
Epoch 1/10
 - 457s - loss: 0.9946 - acc: 0.5002 - val_loss: 0.9612 - val_acc: 0.5000
Epoch 2/10
 - 219s - loss: 0.8776 - acc: 0.5386 - val_loss: 0.8612 - val_acc: 0.7144
Epoch 3/10
 - 265s - loss: 0.7002 - acc: 0.7333 - val_loss: 0.8331 - val_acc: 0.7301
Epoch 4/10
 - 284s - loss: 0.5733 - acc: 0.8754 - val_loss: 0.9227 - val_acc: 0.7129
Train on 9328 samples, validate on 1334 samples
Epoch 1/10
 - 279s - loss: 0.9896 - acc: 0.4992 - val_loss: 0.9377 - val_acc: 0.5000
Epoch 2/10
 - 275s - loss: 0.8633 - acc: 0.6008 - val_loss: 0.8399 - val_acc: 0.7024
Epoch 

In [36]:
print('Average accuracy ', np.mean(np.array(acr)))

Average accuracy  0.7460215556636096


# Conclusion

We can find out from the accuracy of the mixed CNN reaches 0.746, which does not make a much difference when appying words as features alone. However, the speed to train the mixed CNN model is even slower than train on the original CNN. 